src/tnc.rs

```diff
@@ -163,28 +165,86 @@ impl Tnc {
             ConnectConfig::TcpKiss(config) => Box::new(TcpKissTnc::open(&config)?),
             ConnectConfig::LinuxIf(config) => Box::new(LinuxIfTnc::open(&config)?),
         };
-        Ok(Self { imp })
+        Ok(Tnc(Arc::new(Mutex::new(TncInner::new(imp)))))
     }
 
     /// Transmit a frame on the radio. Transmission is not guaranteed even if a
     /// `Ok` result is returned.
     pub fn send_frame(&self, frame: &Ax25Frame) -> Result<(), TncError> {
-        self.imp.send_frame(frame)
+        self.0.lock().unwrap().send_frame(frame)
     }
 
-    /// Block to receive a frame from the radio. If you want to do this on
-    /// a separate thread from sending, clone the `Tnc`.
-    pub fn receive_frame(&self) -> Result<Ax25Frame, TncError> {
-        self.imp.receive_frame()
+    /// Create a new `Receiver<Result<Ax25Frame, TncError>>`
+    /// This will receive a copy of all incoming frames.
+    pub fn incoming(&mut self) -> Receiver<Ax25FrameResult> {
+        self.0.lock().unwrap().incoming()
     }
 }
 
-impl Clone for Tnc {
-    fn clone(&self) -> Self {
-        Tnc {
-            imp: self.imp.clone(),
+pub type Ax25FrameResult = Result<Ax25Frame, Arc<TncError>>;
+
+struct TncInner {
+    imp: Box<dyn TncImpl>,
+    senders: Arc<Mutex<Vec<Sender<Ax25FrameResult>>>>,
+    disconnected: Arc<AtomicBool>,
+}
+
+impl TncInner {
+    fn new(imp: Box<dyn TncImpl>) -> Self {
+        let senders: Arc<Mutex<Vec<Sender<Ax25FrameResult>>>> = Arc::new(Mutex::new(Vec::new()));
+        let disconnected = Arc::new(AtomicBool::new(false));
+
+        {
+            let imp = imp.clone();
+            let senders = senders.clone();
+            let disconnected = disconnected.clone();
+
+            thread::spawn(move || {
+                while !disconnected.load(Ordering::Relaxed) {
+                    let x = match imp.receive_frame() {
+                        Ok(a) => Ok(a),
+                        Err(e) => Err(Arc::new(e)),
+                    };
+
+                    senders.lock().unwrap().retain(|s| {
+                        // If there's an error, remove sender from vec
+                        s.send(x.clone()).is_ok()
+                    });
+                    if x.is_err() {
+                        disconnected.store(true, Ordering::Relaxed);
+                    }
```

Comment:
What is the purpose in persisting the `disconnected` flag? If we instead made it a `loop { }` with a `break` I don't think it would behave any differently.

Separately, I would like to have either a clear explanation why the use of `Relaxed` consistency everywhere is okay, or otherwise replace it with a `Mutex<bool>`, which I find 800% easier to understand.

Reply:
Since `drop` will be changed to shut down `imp`, we can get rid of this boolean entirely.

Reply:
Done.